File: algorithms/typescript/BT-preorder.ts
export type BinaryNode<T> = {
  value: number;
  left: BinaryNode<T>;
  right: BinaryNode<T>;
};
function pre_walk(curr: BinaryNode<number> | null, path: number[]): number[] {
  if (!curr) {
    return path;
  }
  path.push(curr.value);

  pre_walk(curr.left, path);
  pre_walk(curr.right, path);

  // post
  return path;
}
function pre_order_search(head: BinaryNode<number>): number[] {
  return pre_walk(head, []);
}

File: algorithms/python/bt-preorder.py
class BinaryNode:
    def __init__(self) -> None:
        self.value: int
        self.left: BinaryNode
        self.right: BinaryNode


def walk(curr: BinaryNode | None, path: list[int]) -> list[int]:
    if not curr:
        return path
    path.append(curr.value)
    walk(curr.left, path)
    walk(curr.right, path)
    return path


def pre_order_search(head: BinaryNode) -> list[int]:
    return walk(head, [])
